### PR TITLE
8307346 - Add missing gc+phases logging for ObjectCount(AfterGC) JFR event collection code

### PR DESCRIPTION
**Pull request by olivergillespie**
Add logging for the time taken to collect/report ObjectCount(AfterGC) event to Serial, Parallel and G1 full collectors, for parity with the G1 concurrent collector. This event can be *very* expensive to collect (single threaded STW full heap scan), so it's very important to show it clearly to the user.

Added output:
```
Serial

[7.976s][debug][gc,phases      ] GC(7) Trigger cleanups 0.002ms
[7.977s][debug][gc,phases      ] GC(7) Class Unloading 0.556ms
++ [7.977s][debug][gc,phases,start] GC(7) Report Object Count
++ [8.529s][debug][gc,phases      ] GC(7) Report Object Count 552.065ms
[8.529s][info ][gc,phases      ] GC(7) Phase 1: Mark live objects 1066.882ms
[8.529s][info ][gc,phases,start] GC(7) Phase 2: Compute new object addresses

Parallel

[5.786s][debug][gc,phases      ] GC(12) Trigger cleanups 0.002ms
[5.786s][debug][gc,phases      ] GC(12) Class Unloading 0.556ms
++ [5.786s][debug][gc,phases,start] GC(12) Report Object Count
++ [6.313s][debug][gc,phases      ] GC(12) Report Object Count 526.307ms
[6.313s][info ][gc,phases      ] GC(12) Marking Phase 889.900ms
[6.313s][info ][gc,phases,start] GC(12) Summary Phase

G1 Full

[3.922s][debug][gc,phases         ] GC(24) Trigger cleanups 0.002ms
[3.922s][debug][gc,phases         ] GC(24) Phase 1: Class Unloading and Cleanup 0.159ms
++ [3.922s][debug][gc,phases,start   ] GC(24) Report Object Count
++ [4.442s][debug][gc,phases         ] GC(24) Report Object Count 519.292ms
[4.442s][info ][gc,phases         ] GC(24) Phase 1: Mark live objects 653.209ms
[4.442s][info ][gc,phases,start   ] GC(24) Phase 2: Prepare compaction
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307346](https://bugs.openjdk.org/browse/JDK-8307346): Add missing gc+phases logging for ObjectCount(AfterGC) JFR event collection code

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [3ef4f4cc](https://git.openjdk.org/jdk/pull/13772/files/3ef4f4cc36112f4d224a9eb61c53874da6124d93)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13772/head:pull/13772` \
`$ git checkout pull/13772`

Update a local copy of the PR: \
`$ git checkout pull/13772` \
`$ git pull https://git.openjdk.org/jdk.git pull/13772/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13772`

View PR using the GUI difftool: \
`$ git pr show -t 13772`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13772.diff">https://git.openjdk.org/jdk/pull/13772.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13772#issuecomment-1532770562)

### CODE REVIEW

File: src/hotspot/share/gc/parallel/psParallelCompact.cpp [hunk 1]
```diff
@@ -2067,7 +2067,10 @@ void PSParallelCompact::marking_phase(ParallelOldTracer *gc_tracer) {
     JVMCI_ONLY(JVMCI::do_unloading(purged_class));
   }
 
-  _gc_tracer.report_object_count_after_gc(is_alive_closure());
+  {
+    GCTraceTime(Debug, gc, phases) debug("Report Object Count", &_gc_timer);
```
Comment (shipilev) [inline src/hotspot/share/gc/parallel/psParallelCompact.cpp:2071]: Nit: in this file, the holder variables are called `tm`, not `debug`.

File: src/hotspot/share/gc/serial/genMarkSweep.cpp [hunk 2]
```diff
@@ -210,7 +210,10 @@ void GenMarkSweep::mark_sweep_phase1(bool clear_all_softrefs) {
     JVMCI_ONLY(JVMCI::do_unloading(purged_class));
   }
 
-  gc_tracer()->report_object_count_after_gc(&is_alive);
+  {
+    GCTraceTime(Debug, gc, phases) debug("Report Object Count", gc_timer());
```
Comment (shipilev) [inline src/hotspot/share/gc/serial/genMarkSweep.cpp:214]: Nit: in this file, the holder variables are called `tm_m`, not `debug`.

**Review by shipilev — COMMENTED** [review]
Some nits.

**Comment by albertnetymk** [comment]
> single threaded STW full heap scan

`HeapInspection::populate_table` can use multiple threads. Could `report_object_count_after_gc` invoke the parallel version?

**Comment by olivergillespie** [comment]
> Could report_object_count_after_gc invoke the parallel version?

Yes, I was just thinking the same thing! I think it could, I will follow up to implement that change, thanks for the suggestion.

**Comment by shipilev** [comment]
> > Could report_object_count_after_gc invoke the parallel version?
> 
> Yes, I was just thinking the same thing! I think it could, I will follow up to implement that change, thanks for the suggestion.

Filed: https://bugs.openjdk.org/browse/JDK-8307348

**Review by shipilev — APPROVED** [review]
This looks fine to me.

**Review by albertnetymk — APPROVED** [review]

**Pull request closed without merging**